src/ray/raylet/node_manager.cc

```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
 void NodeManager::ProcessRegisterClientRequestMessage(
     const std::shared_ptr<LocalClientConnection> &client, const uint8_t *message_data) {
   auto message = flatbuffers::GetRoot<protocol::RegisterClientRequest>(message_data);
-  client->SetClientID(from_flatbuf<ClientID>(*message->worker_id()));
+  auto client_id = from_flatbuf<ClientID>(*message->worker_id());
+  client->SetClientID(client_id);
   auto worker = std::make_shared<Worker>(message->worker_pid(), message->language(),
                                          message->port(), client);
   if (message->is_worker()) {
     // Register the new worker.
+    bool use_push_task = (worker->Port() > 0);
+    if (use_push_task) {
+      auto entry = worker_clients_.find(client_id);
+      if (entry != worker_clients_.end()) {
+        RAY_LOG(WARNING) << "Received registration of a new worker that already exists: "
+                         << client_id;
+        return;
+      }
+
+      // Initialize a rpc client to the new worker.
+      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
+          new rpc::WorkerTaskClient("0.0.0.0", worker->Port(), client_call_manager_));
```

Comment:
The ip should be `127.0.0.1`?

Reply:
On the server side, we can use `0.0.0.0`, which means the server accepts connection from all IPs. But on the client side, this should be the server IP to connect. I don't understand why this can be `0.0.0.0`. How would the client find the server?

Reply:
while` 0.0.0.0` can work in the test, it's a kind of obscure in the materials I looked at. I will update it to `127.0.0.1` which is more clear in the meaning.